File: dataModule/src/main/java/com/mobnetic/coinguardian/model/market/Quoine.java
package com.mobnetic.coinguardian.model.market;

import java.util.List;

import org.json.JSONArray;
import org.json.JSONObject;

import com.mobnetic.coinguardian.model.CheckerInfo;
import com.mobnetic.coinguardian.model.CurrencyPairInfo;
import com.mobnetic.coinguardian.model.Market;
import com.mobnetic.coinguardian.model.Ticker;

public class Quoine extends Market {

	private final static String NAME = "Quoine";
	private final static String TTS_NAME = NAME;
	private final static String URL = "https://api.quoine.com/products/code/CASH/%1$s";
	private final static String URL_CURRENCY_PAIRS = "https://api.quoine.com/products/";
	
	public Quoine() {
		super(NAME, TTS_NAME, null);
	}

	@Override
	public String getUrl(int requestId, CheckerInfo checkerInfo) {
		return String.format(URL, checkerInfo.getCurrencyPairId());
	}
	
	@Override
	protected void parseTickerFromJsonObject(int requestId, JSONObject jsonObject, Ticker ticker, CheckerInfo checkerInfo) throws Exception {
		ticker.bid = jsonObject.getDouble("market_bid");
		ticker.ask = jsonObject.getDouble("market_ask");
		ticker.vol = jsonObject.getDouble("volume_24h");
		ticker.high = jsonObject.getDouble("high_market_ask");
		ticker.low = jsonObject.getDouble("low_market_bid");
		ticker.last = jsonObject.getDouble("last_traded_price");
	}

	
	// ====================
	// Get currency pairs
	// ====================
	@Override
	public String getCurrencyPairsUrl(int requestId) {
		return URL_CURRENCY_PAIRS;
	}
	
	@Override
	protected void parseCurrencyPairs(int requestId, String responseString, List<CurrencyPairInfo> pairs) throws Exception {
		final JSONArray pairsJsonArray = new JSONArray(responseString);
		for(int i=0; i<pairsJsonArray.length(); ++i) {
			final JSONObject pairJsonObject = pairsJsonArray.getJSONObject(i);
			if (!"CASH".equals(pairJsonObject.getString("code"))) {
				continue;
			}
			
			final String currencyCounter = pairJsonObject.getString("currency");
			final String pairName = pairJsonObject.getString("currency_pair_code");
			if (pairName != null && currencyCounter != null && pairName.endsWith(currencyCounter)) {
				final String currencyBase = pairName.substring(0, pairName.length() - currencyCounter.length());
				pairs.add(new CurrencyPairInfo(
						currencyBase,
						currencyCounter,
						pairName));
			}
		}
	}
}

File: dataModule/src/main/java/com/aneonex/bitcoinchecker/datamodule/model/market/Quoine.kt
package com.aneonex.bitcoinchecker.datamodule.model.market

import com.aneonex.bitcoinchecker.datamodule.model.CheckerInfo
import com.aneonex.bitcoinchecker.datamodule.model.CurrencyPairInfo
import com.aneonex.bitcoinchecker.datamodule.model.Market
import com.aneonex.bitcoinchecker.datamodule.model.Ticker
import org.json.JSONArray
import org.json.JSONObject

class Quoine : Market(NAME, TTS_NAME, null) {
    override fun getUrl(requestId: Int, checkerInfo: CheckerInfo): String {
        return String.format(URL, checkerInfo.currencyPairId)
    }

    @Throws(Exception::class)
    override fun parseTickerFromJsonObject(requestId: Int, jsonObject: JSONObject, ticker: Ticker, checkerInfo: CheckerInfo) {
        ticker.bid = jsonObject.getDouble("market_bid")
        ticker.ask = jsonObject.getDouble("market_ask")
        ticker.vol = jsonObject.getDouble("volume_24h")
        ticker.high = jsonObject.getDouble("high_market_ask")
        ticker.low = jsonObject.getDouble("low_market_bid")
        ticker.last = jsonObject.getDouble("last_traded_price")
    }

    // ====================
    // Get currency pairs
    // ====================
    override fun getCurrencyPairsUrl(requestId: Int): String? {
        return URL_CURRENCY_PAIRS
    }

    @Throws(Exception::class)
    override fun parseCurrencyPairs(requestId: Int, responseString: String, pairs: MutableList<CurrencyPairInfo>) {
        val pairsJsonArray = JSONArray(responseString)
        for (i in 0 until pairsJsonArray.length()) {
            val pairJsonObject = pairsJsonArray.getJSONObject(i)
            if ("CASH" != pairJsonObject.getString("code")) {
                continue
            }
            val currencyCounter = pairJsonObject.getString("currency")
            val pairName = pairJsonObject.getString("currency_pair_code")
            if (pairName.endsWith(currencyCounter)) {
                val currencyBase = pairName.substring(0, pairName.length - currencyCounter.length)
                pairs.add(CurrencyPairInfo(
                        currencyBase,
                        currencyCounter,
                        pairName))
            }
        }
    }

    companion object {
        private const val NAME = "Quoine"
        private const val TTS_NAME = NAME
        private const val URL = "https://api.quoine.com/products/code/CASH/%1\$s"
        private const val URL_CURRENCY_PAIRS = "https://api.quoine.com/products/"
    }
}